problem: can't require peer from outside the actor

JustinAzoff opened the pull request:
This enables an external discovery actor to introduce zyre nodes.

JustinAzoff commented:
I didn't really like adding the forward declaration, but without it you have to reorder a few functions creating a huge diff.. I did that when I first added this feature to my branch, and one of the functions changed as I was testing the feature causing huge merge conflicts.

JustinAzoff commented:
p.s. with this PR this will get master equal to my branch that I've been using to run a network of curve enabled zyre nodes over the internet for the past 2 months!

bluca commented:
zyre_require_peer should go in api/zyre.api then you can regenerate with GSL, so the forward decl is not necessary anymore

JustinAzoff commented:
Ah.. so there are two functions at work here, the new zyre_require_peer api function that I added (that I need to add to the api as you pointed out), and then the existing internal zyre_node_require_peer function that it uses inside the zyre_node actor.

Because of how the functions in zyre_node are ordered, I can't call zyre_node_require_peer from inside zyre_node_recv_api:
```
src/zyre_node.c:619:13: error: implicit declaration of function 'zyre_node_require_peer' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
            zyre_node_require_peer (self, uuid, endpoint, public_key);
```

I added the require_peer to the api.. I think something is off with my gsl/zproject/zproto install though, it's creating a 700 line diff on src/zre_msg.c:

```
 $ git diff --stat
 include/zyre.h              |  72 ++---
 include/zyre_event.h        |  16 +-
 include/zyre_library.h      |  14 +-
 src/zre_msg.c               | 729 ++++++++++++++++++++++++++++++++++++++++--
 src/zre_msg.h               |   8 +
 src/zyre_classes.h          |  19 +-
 src/zyre_private_selftest.c |  14 +-
 src/zyre_selftest.c         |  14 +-
 8 files changed, 789 insertions(+), 97 deletions(-)
```

bluca commented:
Ah ok, I misread the names.

Make sure your GSL version is up to date (packages if  you are running on linux https://build.opensuse.org/package/show/network:messaging:zeromq:git-stable/generator-scripting-language or from the new repo https://github.com/zeromq/gsl )

JustinAzoff commented:
Ah, I was a bit behind on gsl, but it looks like the huge change to zre_msg is from a change in zproto

```
commit 7be75b3e71c8075dd74cf6bbebd67814050e4593
Author: Michal Vyskocil <michal.vyskocil@gmail.com>
Date:   Thu Jun 8 15:43:26 2017 +0200

    Problem: want to store codec_c messages as ZPL/zconfig

    Solution: port the code from codec_v1 - I know it does bloat API a bit, however
    this is complicated code, which must be generated in order to make it correct
```

Which confuses me because other people have regen'ed lately and not caused this to be updated, so I'm not sure if something is wrong on my end that is pulling in this big diff.  I updated the header that needed to be changed, but left this as it was for now:

```
 src/zre_msg.bnf |  14 +-
 src/zre_msg.c   | 743 ++++++++++++++++++++++++++--
 src/zre_msg.h   |  22 +-
 3 files changed, 733 insertions(+), 46 deletions(-)
```

bluca commented:
Right, you can avoid regenerating that for now, just run gsl project.xml
